File: clients/typescript/src/lipi/varint.ts
import { Bytes } from "../utils/bytes.ts";

export function encodeVarInt(num: bigint | number) {
    num = BigInt(num);
    if (num < 0n) throw new RangeError(`expected unsigned number: found ${num}`);

    let buf = [];
    while (num > 0b111_1111) {
        buf.push(Number((num & 0xFFn) | 0b1000_0000n));
        num >>= 7n;
    }
    buf.push(Number(num));
    return new Uint8Array(buf);
}

export function decodeVarInt(bytes: Bytes): bigint {
    let result = 0n;
    let shift = 0n;

    while (true) {
        let byte = BigInt(bytes.nextByte());
        if (shift == 63n && byte >= 2) throw new Error("invalid variable-length integer");

        if ((byte & 0b1000_0000n) == 0n) return result | (byte << shift);

        result |= (byte & 0b111_1111n) << shift; // low-order 7 bits of value
        shift += 7n;
    }
}

File: clients/dart/lib/lipi/varint.dart
import 'dart:typed_data';

import 'package:setu_client/utils/bytes.dart';
import 'package:setu_client/utils/common.dart';

Uint8List encodeVarInt(BigInt num) {
  assertExpr(
    num >= BigInt.zero,
    error: RangeError.new,
    message: () => 'expected unsigned number: found $num',
  );

  final buf = <int>[];

  while (num > BigInt.from(0x7F)) {
    buf.add(((num & BigInt.from(0xFF)).toInt()) | 0x80);
    num = num >> 7;
  }

  buf.add(num.toInt());

  return Uint8List.fromList(buf);
}

BigInt decodeVarInt(Bytes bytes) {
  var result = BigInt.zero;
  var shift = 0;

  while (true) {
    final byte = BigInt.from(bytes.nextByte());

    if (shift == 63 && byte >= BigInt.two) {
      throw Exception('invalid variable-length integer');
    }

    if ((byte & BigInt.from(0x80)) == BigInt.zero) {
      return result | (byte << shift);
    }

    result |= (byte & BigInt.from(0x7F)) << shift;
    shift += 7;
  }
}
